cdc/sinkv2/util/mq/kafka_producer_helper.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mq
+
+import (
+	"net/url"
+	"strings"
+
+	"github.com/Shopify/sarama"
+	"github.com/pingcap/tiflow/cdc/sink/mq/codec"
+	"github.com/pingcap/tiflow/cdc/sink/mq/manager"
+	"github.com/pingcap/tiflow/cdc/sink/mq/producer/kafka"
+	"github.com/pingcap/tiflow/pkg/config"
+	cerror "github.com/pingcap/tiflow/pkg/errors"
+	pkafka "github.com/pingcap/tiflow/pkg/kafka"
+)
+
+// TODO: add tests for below funcs.
+
+// GetTopic returns the topic name from the sink URI.
+func GetTopic(sinkURI *url.URL) (string, error) {
+	topic := strings.TrimFunc(sinkURI.Path, func(r rune) bool {
+		return r == '/'
+	})
+	if topic == "" {
+		return "", cerror.ErrKafkaInvalidConfig.GenWithStack("no topic is specified in sink-uri")
+	}
+	return topic, nil
+}
+
+// GetProtocol returns the protocol from the sink URI.
+func GetProtocol(protocolStr string) (config.Protocol, error) {
+	var protocol config.Protocol
+	if err := protocol.FromString(protocolStr); err != nil {
+		return protocol, cerror.WrapError(cerror.ErrKafkaInvalidConfig, err)
+	}
+
+	return protocol, nil
+}
+
+// GetEncoderConfig returns the encoder config and validates the config.
+func GetEncoderConfig(
+	sinkURI *url.URL,
+	protocol config.Protocol,
+	replicaConfig *config.ReplicaConfig,
+	maxMsgBytes int,
+) (*codec.Config, error) {
+	encoderConfig := codec.NewConfig(protocol)
+	if err := encoderConfig.Apply(sinkURI, replicaConfig); err != nil {
+		return nil, cerror.WrapError(cerror.ErrKafkaInvalidConfig, err)
+	}
+	// Always set encoder's `MaxMessageBytes` equal to producer's `MaxMessageBytes`
+	// to prevent that the encoder generate batched message too large
+	// then cause producer meet `message too large`.
+	encoderConfig = encoderConfig.WithMaxMessageBytes(maxMsgBytes)
+
+	if err := encoderConfig.Validate(); err != nil {
+		return nil, cerror.WrapError(cerror.ErrKafkaInvalidConfig, err)
+	}
+
+	return encoderConfig, nil
+}
+
+// GetTopicManagerAndTryCreateTopic returns the topic manager and try to create the topic.
+func GetTopicManagerAndTryCreateTopic(
+	endpoints []string,
+	topic string,
+	topicCfg *kafka.AutoCreateTopicConfig,
+	adminClient pkafka.ClusterAdminClient,
+	saramaConfig *sarama.Config,
+) (manager.TopicManager, error) {
+	client, err := sarama.NewClient(endpoints, saramaConfig)
```

Comment:
Can `sarama.NewClient` block? If so, should we create it asynchronously?

Reply:
It is possible, but we have to hold it to create the sync producer.

Reply:
Removed, we can pass a client from the caller.

Reply:
Good catch. It was a mistake when I refactor the code.

Reply:
`sarama.NewClient` would be blocked, but we have to hold it to create the producer. The producer needs to hold it to close itself quickly.